src/Compilers/CSharp/Portable/Parser/LanguageParser.cs

```diff
@@ -7164,12 +7164,54 @@ bool canFollowNullableType()
+                        // These contexts might be a type that is at the end of an expression.
+                        // For backward compatibility we want to consider a `?` token as part of the `?:` operator if possible.
+                        // However, if current token is an identifier, it might be beneficial to allow `?`
+                        // and treat this identifier as a designation. Nullable types in patterns are semantically invalid,
+                        // so we will get a nice error about that during binding
+                        if (this.CurrentToken.Kind == SyntaxKind.IdentifierToken)
+                        {
+                            // Given that we are deciding whether we take `?:` operator or not while looking at state after `?`,
+                            // our focus is on `:` token as well. But `:` have special meaning in context of switch statements
+                            // and expressions (since `:` in switch expressions is recovered to `=>`). So consider the following case:
+                            // `case T? t:`. On its own `T ? t :` can be treated as a conditional expression with condition `T`,
+                            // 'whenTrue' `t` and 'whenFalse' not yet parsed. But it is way better to take `:` as case label end and
+                            // thus parse label condition as a declaration pattern `T? t`. `whenIsKeyword` flag can conveniently tell
+                            // whether we are in a 'top-level' state of a label, so if condition get parenthesized like `case (a ? b : c):`
+                            // we don't take the first `:` as label end and parse label condition as a parenthesized conditional expression
+                            if ((_termState.HasFlag(TerminatorState.IsExpressionOrPatternInCaseLabelOfSwitchStatement) ||
+                                 _termState.HasFlag(TerminatorState.IsPatternInSwitchExpressionArm)) && whenIsKeyword)
+                            {
+                                return true;
+                            }
+
+                            // Special case for `await`: if current token is an `await` identifier and the next token
+                            // can start an expression, prefer parsing conditional expression
+                            // rather than having `await` as a pattern designation
+                            if (this.CurrentToken.ContextualKind == SyntaxKind.AwaitKeyword)
+                            {
+                                using var _ = GetDisposableResetPoint(resetOnDispose: true);
+                                this.EatToken();
+                                return !CanStartExpression();
+                            }
+
+                            var nextTokenKind = PeekToken(1).Kind;
+
+                            // These token either 100% end a pattern or start a new one
+                            return SyntaxFacts.IsLiteral(nextTokenKind) ||
```

Comment:
i don't understand the literal case.  That would something like `x is Y? a 4`?  Confused by that.

Also, could you give exmples to each of these cases to indicate why it's felt to be ok that this ends/starts a pattern.

Reply:
i also don't understand the ISPRedefinedTYpe check for the same reason.

Reply:
Added explanation comment about every token/token type in that list. For literals and predefined types added more tests to show the intended behavior